### PR TITLE
Add centernet

### PR DESCRIPTION
**Pull request by jerrywgz**

### CODE REVIEW

File: ppdet/modeling/necks/centernet_fpn.py [hunk 1]
```diff
@@ -141,18 +148,21 @@ class CenterNetDLAFPN(nn.Layer):
         first_level (int): the first level of input feature fed into the upsamplng
```
Comment (nemonameless) [inline ppdet/modeling/necks/centernet_fpn.py]: ‘first_level’  conflict

File: ppdet/data/transform/operators.py [hunk 2]
```diff
@@ -2962,3 +2963,264 @@ def apply(self, sample, context=None):
...
+        ]
+        distortions = np.random.permutation(functions)
+        for func in distortions:
+            img = func(img, img_gray)
+        sample['image'] = img
+        return sample
+
+
+@register_op
+class Gt2CenterTarget(BaseOperator):
```
Comment (nemonameless) [inline ppdet/data/transform/operators.py]: 建议起名为Gt2CenterNetTarget，且放在batch_operators.py里

File: ppdet/modeling/necks/centernet_fpn.py (hunk 1 above)
Comment (jerrywgz) [reply ppdet/modeling/necks/centernet_fpn.py]: done

File: ppdet/data/transform/operators.py (hunk 2 above)
Comment (jerrywgz) [reply ppdet/data/transform/operators.py]: done，thx

File: configs/centernet/README.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+## Model Zoo
+
+### CenterNet Results on COCO-val 2017
+
+| backbone       | input shape | mAP   |    FPS    | download | config |
+| :--------------| :------- |  :----: | :------: | :----: |:-----: |
+| DLA-34(paper)  | 512x512 |  37.4  |     -   |    -   |   -    |
+| DLA-34         | 512x512 |  37.6  |     -   | [model](https://bj.bcebos.com/v1/paddledet/models/centernet_dla34_1x.pdparams) | [config](https://github.com/PaddlePaddle/PaddleDetection/tree/develop/configs/centernet/centernet_dla34_1x.yml) |
+| ResNet50 + DLAUp  | 512x512 |  38.9  |     -   | [model](https://bj.bcebos.com/v1/paddledet/models/centernet_r50_1x.pdparams) | [config](https://github.com/PaddlePaddle/PaddleDetection/tree/develop/configs/centernet/centernet_r50_1x.yml) |
```
Comment (FlyingQianMM) [inline configs/centernet/README.md]: config文件可以写成相对路径./centernet_r50_1x.yml

File: deploy/python/preprocess.py [hunk 4]
```diff
@@ -245,6 +245,139 @@ def __call__(self, im, im_info):
...
+                             input_size,
+                             rot,
+                             output_size,
+                             shift=(0., 0.),
+                             inv=False):
+        """Get the affine transform matrix, given the center/scale/rot/output_size.
+
+        Args:
+        center (np.ndarray[2, ]): Center of the bounding box (x, y).
+        scale (np.ndarray[2, ]): Scale of the bounding box
```
Comment (FlyingQianMM) [inline deploy/python/preprocess.py]: scale没有对应的变量

File: deploy/python/preprocess.py [hunk 5]
```diff
@@ -245,6 +245,139 @@ def __call__(self, im, im_info):
...
+                 shift=0.1):
+        self.keep_res = keep_res
+        self.pad = pad
+        self.input_h = input_h
+        self.input_w = input_w
+        self.scale = scale
+        self.shift = shift
+
+    def _get_3rd_point(self, a, b):
+        assert len(a) == 2
```
Comment (FlyingQianMM) [inline deploy/python/preprocess.py]: 给出报错信息提示？

File: deploy/python/preprocess.py [hunk 6]
```diff
@@ -245,6 +245,139 @@ def __call__(self, im, im_info):
...
+
+        Returns:
+        np.ndarray: The transform matrix.
+        """
+        assert len(center) == 2
+        assert len(output_size) == 2
+        assert len(shift) == 2
+
+        if not isinstance(input_size, np.ndarray) and not isinstance(input_size,
+                                                                     list):
```
Comment (FlyingQianMM) [inline deploy/python/preprocess.py]: not isinstance(input_size, (np.ndarray, list))

File: ppdet/data/transform/batch_operators.py [hunk 7]
```diff
@@ -25,18 +25,26 @@
...
 import numpy as np
 from .operators import register_op, BaseOperator, Resize
-from .op_helper import jaccard_overlap, gaussian2D
+from .op_helper import jaccard_overlap, gaussian2D, gaussian_radius, draw_umich_gaussian
 from .atss_assigner import ATSSAssigner
 from scipy import ndimage
 
 from ppdet.modeling import bbox_utils
 from ppdet.utils.logger import setup_logger
+from ...modeling.keypoint_utils import get_affine_transform, affine_transform
```
Comment (FlyingQianMM) [inline ppdet/data/transform/batch_operators.py]: from ppdet.modeling.

File: ppdet/modeling/keypoint_utils.py [hunk 8]
```diff
@@ -61,10 +61,12 @@ def get_affine_transform(center,
         np.ndarray: The transform matrix.
     """
     assert len(center) == 2
-    assert len(input_size) == 2
     assert len(output_size) == 2
     assert len(shift) == 2
 
+    if not isinstance(input_size, np.ndarray) and not isinstance(input_size,
```
Comment (FlyingQianMM) [inline ppdet/modeling/keypoint_utils.py]: isinstance(input_size, (np.ndarray, list))

File: ppdet/modeling/heads/centernet_head.py [hunk 9]
```diff
@@ -16,7 +16,7 @@
 import paddle
 import paddle.nn as nn
 import paddle.nn.functional as F
-from paddle.nn.initializer import KaimingUniform
+from paddle.nn.initializer import KaimingUniform, Constant, Uniform
```
Comment (FlyingQianMM) [inline ppdet/modeling/heads/centernet_head.py]: KaimingUniform没有用到

File: ppdet/modeling/necks/centernet_fpn.py [hunk 10]
```diff
@@ -16,8 +16,9 @@
 import math
 import paddle
 import paddle.nn as nn
+from paddle import ParamAttr
+from paddle.nn.initializer import KaimingUniform, Uniform
```
Comment (FlyingQianMM) [inline ppdet/modeling/necks/centernet_fpn.py]: del KaimingUniform

File: configs/centernet/README.md (hunk 3 above)
Comment (jerrywgz) [reply configs/centernet/README.md]: done, thx

File: deploy/python/preprocess.py (hunk 5 above)
Comment (jerrywgz) [reply deploy/python/preprocess.py]: done, thx

File: deploy/python/preprocess.py (hunk 4 above)
Comment (jerrywgz) [reply deploy/python/preprocess.py]: done, thx

File: deploy/python/preprocess.py (hunk 6 above)
Comment (jerrywgz) [reply deploy/python/preprocess.py]: done, thx

File: ppdet/data/transform/batch_operators.py (hunk 7 above)
Comment (jerrywgz) [reply ppdet/data/transform/batch_operators.py]: done, thx

File: ppdet/modeling/heads/centernet_head.py (hunk 9 above)
Comment (jerrywgz) [reply ppdet/modeling/heads/centernet_head.py]: done, thx

File: ppdet/modeling/necks/centernet_fpn.py (hunk 10 above)
Comment (jerrywgz) [reply ppdet/modeling/necks/centernet_fpn.py]: done, thx

File: configs/centernet/centernet_dla34_140e_coco.yml [hunk 11]
```diff
@@ -0,0 +1,9 @@
+_BASE_: [
+  '../datasets/coco_detection.yml',
+  '../runtime.yml',
+  '_base_/optimizer_140e.yml',
+  '_base_/centernet_dla34.yml',
+  '_base_/centernet_reader.yml',
+]
+
+weights: output/centernet_dla34_1x/model_final
```
Comment (nemonameless) [inline configs/centernet/centernet_dla34_140e_coco.yml]: centernet_dla34_1x -> centernet_dla34_140e_coco

File: configs/centernet/centernet_r50_140e_coco.yml [hunk 12]
```diff
@@ -0,0 +1,9 @@
+_BASE_: [
+  '../datasets/coco_detection.yml',
+  '../runtime.yml',
+  '_base_/optimizer_140e.yml',
+  '_base_/centernet_r50.yml',
+  '_base_/centernet_reader.yml',
+]
+
+weights: output/centernet_r50_1x/model_final
```
Comment (nemonameless) [inline configs/centernet/centernet_r50_140e_coco.yml]: centernet_r50_1x -> centernet_r50_140e_coco

File: configs/centernet/README_cn.md [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+[CenterNet](http://arxiv.org/abs/1904.07850)是Anchor Free检测器，将物体表示为一个目标框中心点。CenterNet使用关键点检测的方式定位中心点并回归物体的其他属性。CenterNet是以中心点为基础的检测方法，是端到端可训练的，并且相较于基于anchor的检测器更加检测高效。
+
+## 模型库
+
+### CenterNet在COCO-val 2017上结果
+
+| 骨干网络       | 输入尺寸 | mAP   |    FPS    | 下载链接 | 配置文件 |
+| :--------------| :------- |  :----: | :------: | :----: |:-----: |
+| DLA-34(paper)  | 512x512 |  37.4  |     -   |    -   |   -    |
+| DLA-34         | 512x512 |  37.6  |     -   | [下载链接](https://bj.bcebos.com/v1/paddledet/models/centernet_dla34_140e.pdparams) | [配置文件](https://github.com/PaddlePaddle/PaddleDetection/tree/develop/configs/centernet/centernet_dla34_1x.yml) |
```
Comment (nemonameless) [inline configs/centernet/README_cn.md]: download:   https://paddledet.bj.bcebos.com/models/centernet_dla34_140e_coco.pdparams,      
config:   ./centernet_dla34_140e_coco.yml

File: configs/centernet/centernet_dla34_140e_coco.yml (hunk 11 above)
Comment (jerrywgz) [reply configs/centernet/centernet_dla34_140e_coco.yml]: done

File: configs/centernet/centernet_r50_140e_coco.yml (hunk 12 above)
Comment (jerrywgz) [reply configs/centernet/centernet_r50_140e_coco.yml]: done

File: configs/centernet/README_cn.md (hunk 13 above)
Comment (jerrywgz) [reply configs/centernet/README_cn.md]: done

**Review by nemonameless — APPROVED** [review]

**Review by FlyingQianMM — APPROVED** [review]

**Pull request merged**